src/vs/platform/actionWidget/browser/actionWidget.ts

```diff
@@ -0,0 +1,634 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import { ActionBar } from 'vs/base/browser/ui/actionbar/actionbar';
+import * as dom from 'vs/base/browser/dom';
+import { IAnchor } from 'vs/base/browser/ui/contextview/contextview';
+import { KeybindingLabel } from 'vs/base/browser/ui/keybindingLabel/keybindingLabel';
+import { IListEvent, IListMouseEvent, IListRenderer, IListVirtualDelegate } from 'vs/base/browser/ui/list/list';
+import { List } from 'vs/base/browser/ui/list/listWidget';
+import { IAction } from 'vs/base/common/actions';
+import { Disposable, DisposableStore, IDisposable, MutableDisposable } from 'vs/base/common/lifecycle';
+import { OS } from 'vs/base/common/platform';
+import { localize } from 'vs/nls';
+import { ICommandService } from 'vs/platform/commands/common/commands';
+import { IContextKeyService, RawContextKey } from 'vs/platform/contextkey/common/contextkey';
+import { IContextViewService } from 'vs/platform/contextview/browser/contextView';
+import { IKeybindingService } from 'vs/platform/keybinding/common/keybinding';
+import { ITelemetryService } from 'vs/platform/telemetry/common/telemetry';
+import { Codicon } from 'vs/base/common/codicons';
+import { ActionSet } from 'vs/base/common/actionWidget/actionWidget';
+import 'vs/css!./actionWidget';
+import { createDecorator, ServicesAccessor } from 'vs/platform/instantiation/common/instantiation';
+import { InstantiationType, registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { Action2, registerAction2 } from 'vs/platform/actions/common/actions';
+import { KeyCode, KeyMod } from 'vs/base/common/keyCodes';
+import { KeybindingWeight } from 'vs/platform/keybinding/common/keybindingsRegistry';
+
+export const Context = {
+	Visible: new RawContextKey<boolean>('codeActionMenuVisible', false, localize('codeActionMenuVisible', "Whether the code action list widget is visible"))
+};
+export interface IRenderDelegate<T> {
+	onHide(cancelled: boolean): void;
+	onSelect(action: T, trigger: any, preview?: boolean): Promise<any>;
+}
+
+export interface ActionShowOptions {
+	readonly includeDisabledActions: boolean;
+	readonly fromLightbulb?: boolean;
+	readonly showHeaders?: boolean;
+}
+
+export interface ListMenuItem<T> {
+	item?: T;
+	kind?: any;
+	group?: any;
+	disabled?: boolean | string;
```

Comment:
I'm not sure what disabled as a string means

Reply:
the code action provides a message about why it's disabled... could use a better name for sure

Reply:
That was my guess but wasn't sure, could we add a /** comment */ above so it's clear to consumers?